src/merkletree.ts

```diff
@@ -0,0 +1,166 @@
+import { BytesLike, HexString, compare } from './bytes';
+
+import {
+  MultiProof,
+  makeMerkleTree,
+  isValidMerkleTree,
+  getProof,
+  getMultiProof,
+  processProof,
+  processMultiProof,
+  renderMerkleTree,
+} from './core';
+
+import { MerkleTreeOptions, defaultOptions } from './options';
+import { validateArgument, invariant } from './utils/errors';
+
+export type MerkleTreeData<T> = {
+  format: string;
+  tree: HexString[];
+  values: { value: T; treeIndex: number }[];
+};
+
+export interface MerkleTree<T> {
```

Comment:
Ideally we would have a way to prevent T from being `number`, as this leads to confusion between leaf value and leaf index.

I tried doing:
```typescript
export abstract class MerkleTreeImpl<T, _ = Exclude<T, number>> implements MerkleTree<T> {
```
But it doesn't work as I intended.

Reply:
Couldn't make it work. I think the best we can do is restrict it at the MerkleTree constructor.

I added an update, it'd be the last thing imo, otherwise LGTM.

Reply:
LGTM